cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java

```diff
@@ -18,221 +18,73 @@
- * </li>
- * </ol>
+ * The first command line argument to this class is the name of the user's Spark program class.
  */
-
 public class SparkProgramWrapper {
 
   private static final Logger LOG = LoggerFactory.getLogger(SparkProgramWrapper.class);
-  private static final int PROGRAM_WRAPPER_ARGUMENTS_SIZE = 1;
-  private final String[] arguments;
-  private final Class<? extends SparkProgram> userProgramClass;
-  private static BasicSparkContext basicSparkContext;
-  private static SparkContext sparkContext;
-  private static boolean scalaProgram;
 
-  // TODO: Get around Spark's limitation of only one SparkContext in a JVM and support multiple spark context:
-  // CDAP-4
-  private static boolean sparkProgramSuccessful;
-  private static boolean sparkProgramRunning;
+  public static void main(String[] args) throws Exception {
+    new SparkProgramWrapper(args);
+  }
 
   /**
-   * Constructor
+   * Constructor. Starts a new {@link org.apache.spark.SparkContext} and exeuctes the user {@link SparkProgram}.
    *
    * @param args the command line arguments
    * @throws RuntimeException if the user's program class is not found
    */
   @SuppressWarnings("unchecked")
-  private SparkProgramWrapper(String[] args) {
-    arguments = validateArgs(args);
-    try {
-      // Load the user class from the ProgramClassLoader
-      userProgramClass = loadUserSparkClass(arguments[0]);
-    } catch (ClassNotFoundException e) {
-      LOG.error("Unable to load program class: {}", arguments[0], e);
-      throw Throwables.propagate(e);
-    }
-  }
+  private SparkProgramWrapper(String[] args) throws Exception {
+    Preconditions.checkArgument(args.length >= 1, "Missing SparkProgram class name.");
 
-  public static void main(String[] args) {
-    new SparkProgramWrapper(args).runUserProgram();
-  }
+    // Load the user class from the ProgramClassLoader
+    Class<? extends SparkProgram> sparkProgramClass = loadUserSparkClass(args[0]);
+    ExecutionSparkContext context = setupSparkContext(sparkProgramClass);
+    LOG.debug("Launching user spark program {} from class {}", context, sparkProgramClass.getName());
 
-  /**
-   * Validates command line arguments being passed
-   * Expects at least {@link SparkProgramWrapper#PROGRAM_WRAPPER_ARGUMENTS_SIZE} command line arguments to be present
-   *
-   * @param arguments String[] the arguments
-   * @return String[] if the command line arguments are sufficient else throws a {@link RuntimeException}
-   * @throws IllegalArgumentException if the required numbers of command line arguments were not present
-   */
-  private String[] validateArgs(String[] arguments) {
-    if (arguments.length < PROGRAM_WRAPPER_ARGUMENTS_SIZE) {
-      throw new IllegalArgumentException("Insufficient number of arguments. Program class name followed by its" +
-                                           " arguments (if any) should be provided");
-    }
-    return arguments;
+    sparkProgramClass.newInstance().run(context);
```

Comment:
I also thought about whether it should be like the current way or like this in the `main()` method.

```java
new SparkProgramWrapper(args).run();
```

Any preferences?